RawKV Change Data Capture (#86)
tikv/rfcs: text/0083-rawkv-cross-cluster-replication.md

# RawKV Cross Cluster Replication

- RFC PR: https://github.com/tikv/rfcs/pull/86
- Tracking Issue: https://github.com/tikv/repo/issues/0000

## Summary

This proposal introduces the technical design of RawKV Cross Cluster Replication.

## Motivation

Customers are deploying TiKV clusters as non-transaction Key-Value (RawKV) storage. But the lack of *Cross Cluster Replication* is a obstacle to deploy disaster recovery clusters for providing more highly available services.

> The ability of cross cluster replication will help TiKV be more adoptable to industries such as Banking and Finance.

Reviewer: Could you list key challenges of the use case "Banking" plus "Cross Cluster Replication"? It would be helpful for understanding the design in the RFC.

Here is my initial thought,
* How to address high latency issue between clusters?
* How do we define the consistency of raw kv?

Author: @overvenus I add a chapter about "Key Challenges & Solutions", and specify that we will implement RawKV features in TiCDC as new modules for isolation with original TiDB related logics.
PTAL, thanks~

# RawKV Cross Cluster Replication

- RFC PR: https://github.com/tikv/rfcs/pull/86
- Tracking Issue: https://github.com/tikv/repo/issues/0000

## Summary

This proposal introduces the technical design of RawKV Cross Cluster Replication.

## Motivation

Customers are deploying TiKV clusters as non-transaction Key-Value (RawKV) storage. But the lack of *Cross Cluster Replication* is a obstacle to deploy disaster recovery clusters for providing more highly available services.

The ability of cross cluster replication will help TiKV to be more adoptable to industries such as Banking and Finance.

## Key Challenges & Solutions

#### Capturing Change

> As there is no timestamp in RawKV entries, it is not possible to capture which and when data is changed. So in this proposal, we add timestamp to data. See "2. Add Timestamp to Data" chapter.

Reviewer 2: ```suggestion
As there is no timestamp in RawKV entries, it is not possible to capture which and when data is changed. So in this proposal, we add timestamp to data. See "(Add Timestamp to Data)[#2. Add Timestamp to Data]" chapter.
```

> #### Consistence

Reviewer 2: ```suggestion
#### Consistency
```

In this proposal, we provide a TiKV sink with configurable concurrency, batch size, and backoff parameters, to be adapted to different network environments.

## Detailed Design

### 1. Utilize TiKV-BR & TiKV-CDC as Data Replication Component

**TiKV-BR**, a fork of [BR](https://github.com/pingcap/tidb/tree/master/br), is the tool for backup and restoration of TiKV. We utilize **TIKV-BR** to accomplish data initialization of recovery cluster.

**TiKV-CDC**, a fork of [TiCDC](https://github.com/pingcap/tiflow), is the tool for replicating incremental data of TiKV, which gets change data from TiKV's observer component, with high performance, high available, and scalable.

Besides, **TiKV-CDC** can also provide the ability to connect to other components, e.g. Message Queues, which will help TiKV integrate with customers' data systems, and further extend the applicable scenarios of TiKV.

![rawkv-cdc](../media/rawkv-cdc.png)

### 2. Add Timestamp to Data

> As a kind of [Change Data Capture](https://en.wikipedia.org/wiki/Change_data_capture), timestamp (or version) is necessary to indicate which data is changed, and when. While RawKV doesn't have such a thing before, we need to add timestamp to the RawKV data.

Reviewer 2: ```suggestion
As a kind of [Change Data Capture](https://en.wikipedia.org/wiki/Change_data_capture), timestamp (or version) is necessary to indicate which and when data is changed.
```

#### 2.1 Requirement

> Among requests of a key, *Order of Timestamp* must be consistent with sequence of data flush to disk in TiKV.

Reviewer 2: ```suggestion
Among requests of a key, the timestamp must be monotonic with the sequence of data flushed to disk in TiKV.
```

At the same time, as RawKV doesn't provide cross-rows transaction and snapshot isolation, we treat requests of different keys as concurrent, to reduce complexity and improve performance.

#### 2.2 Timestamp Generation

Timestamp is generated by TiKV internally, to get a better overall performance and client compatibility.

We use [HLC](https://cse.buffalo.edu/tech-reports/2014-04.pdf) method to generate timestamp, which not only is easy to use (by being closed to real world time), but also can capture causality.

##### 2.2.1 Physical Part of Timestamp

Physical part of HLC is acquired from TSO of [PD](https://github.com/tikv/pd).

TSO is a global monotonically increasing timestamp, which help the generation of timestamp be independent to local clock of machine, and be immune to issues such as reverse between machine reboot.

> Physical part is refreshed by repeatedly acquiring TSO in a period of `500ms`, to keep it being closed the real world time. And it can tolerate fault of TSO no longer than `30s`, to keep time-related metrics such as RPO reasonable.

Reviewer 2: Why `30s`? And will raw write fails if TSO is offline over 30s?

Author: En... I think this is not a serious decision. Seems that offline of TSO doesn't break the correctness, except some time-related metrics.
Maybe we can drop this restriction, and report failure metrics. And add some tests on this scenario.

Meanwhile, TiKV-CDC is highly dependent to PD. When PD is offline, the replication is not working.

Reviewer 2: > TiKV-CDC is highly dependent to PD. When PD is offline, the replication is not working. 

Yes, but the write on the main TiKV may still work.

> Maybe we can drop this restriction, and report failure metrics.

Agree

Author: Done, PTAL~

> Besides, on startup, physical part must be initialized by a successful TSO.

Reviewer 3: what happens if it failed to acquire ts from PD?

Author: The raw write operations (raw_put, raw_del, etc.) will fail.

Reviewer 2: Even without Raw CDC, TiKV can't startup when PD is out of service.

Reviewer 3: ok. it would be better to add iit to the RFC for disambiguation.

Author: Done, PTAL~

##### 2.2.2 Logical Part of Timestamp

Logical part is advanced on every write request.

> Moreover, logical part is advanced on *Leader Transfer*. As TiKV a distributed system, every store in TiKV cluster has a instance of timestamp generator. As every key is existed in only one region and only one peer (the leader) in a region can write, it's safe to generate timestamp locally in each store, except for leader transfer.

Reviewer 2: ```suggestion
Moreover, the logical part is advanced on *Leader Transfer*. As TiKV is a distributed system, every store in the TiKV cluster has an instance of timestamp generator. As every key exists in only one region and only one peer (the leader) in the region can write, it's safe to generate timestamps locally in each store, except for leader transfer.
```

> On leader transfer, the timestamp generated by other store would be bigger than the store where new leader located, which will violate the causal consistency requirement. We solve this by this method: Every peer observes raft message applied, and keep the maximum timestamp of the region (called `Region-Max-TS`). On become leader, this peer advances timestamp generator of store to be bigger than `Region-Max-TS`. 

Reviewer 2: ```suggestion
On leader transfer, the timestamp generated by another store could be larger than the store where the new leader is located, which will violate the causal consistency requirement. We solve this by this method: Every peer observes the raft message applied and keeps the maximum timestamp of the region (called `Region-Max-TS`). To become leader, this peer advances the timestamp generator of the store to be bigger than `Region-Max-TS`. 
```

Besides, other region changes should be carefully handled:

> * On region merge: The `Region-Max-TS` of the region is the bigger one of the original two regions.

Reviewer 2: ```suggestion
* On region merge: The `Region-Max-TS` of the region is the max of the two original regions.
```

**TiKV-CDC** using *Resolved Timestamp* (or `resolved-ts`) to indicate the largest transaction timestamp of the replication task, to help downstream avoid getting partial data of a transaction. Downstream streaming system can also use `resolved-ts` as [*watermark*](https://www.oreilly.com/radar/the-world-beyond-batch-streaming-102/).

As no transaction support, in RawKV scenario we don't need scanning locks in *lock_cf*. But we should notify `resolved-ts` event generation that the minimum timestamp of data that on the way from *proposed* to *applied* (by [`resolver.track_lock`](https://github.com/tikv/tikv/blob/v5.0.4-20211201/components/resolved_ts/src/resolver.rs#L47-L57)).

### 3. Deletion

To make deletions be captured as change data, we turn physical deletion to logical deletion, i.e., set a *deleted* flag, other than physically delete the entry.

The *garbage collection* of deleted data is implemented by setting TTL to 25 hours after, and *GC-ed* by compaction filter. At the same time, start timestamp (i.e. `start-ts`) of **TiKV-CDC** replication task should not be earlier than 24 hours before.

### 4. Encoding

Timestamp & deleted flag are encoded as following:

```

> r{keyspace id}{user key}{MCE Padding}{^timestamp:u64}: {user value}{expire ts}{meta flags}

Reviewer 3: should the `keyspace id` part be removed?

Author: No. Why it should be removed ?

Reviewer 2: Ah, it's because we've decided not to include keyspace in API V2.

Author: Er... OK, so what's the latest proposal of API V2 ?

* Timestamp encoded into key will be beneficial to the following, comparing with encoded into value:
  * More effective catch up, as we can get scope of change data without read & decode all values.
  * [PiTR](https://en.wikipedia.org/wiki/Point-in-time_recovery) is available.
  * Less code complexity, as data structure of RawKV & TxnKV are the same.
* And with the following acceptable loss:
  * More read latency in tens of microseconds.
  * A little more write overhead caused by [MCE Padding](https://github.com/facebook/mysql-5.6/wiki/MyRocks-record-format#memcomparable-format).
  * An extra Garbage Collection is required (used with TxnKV).
  * Possible more space occupied (related to update pattern).

### 5. Incremental Scan

On replication task resumed after paused, we scan *KVDB* to get the change data during pause by extracting timestamp encoded in key. The scan will be heavy as we actually scan all the keys, so we should avoid scanning as possible, and limit scan rate.

> *Incremental Scan in SQL / TxnKV scenario scans write_cf to get change data scope, which is much cheaper. So an alternative to reduce cost for RawKV is writing timestamp to write_cf other than encoded in key, but will introduce an additional write for each put request. We tend to the lower delay of writing now.*

Reviewer 3: what's the solution for writing `timestamp`?

Author: It would be: in addition to `{key}:{value}` in default_cf, write an entry `{key}:{timestamp}` to write_cf, to indicated the latest change timestamp of the key.

  
  * Key range splitting is align with boundary of Regions. For simplicity, each *sub-range* contains almost equal number of Regions.
  
  * *Rebalance* can only be performed by *tikv-cdc cli*. As *rebalance* will trigger *incremental scan* and impact performance of TiKV, it is not running automatically. We need more information to determine the best moment for *rebalance* by TiKV-CDC itself in the future.

* Capture: Accepts key-range tasks, and creates *KV Processors* to handle them.

* Processor:
  
  * A new *KV Processor* which accepts key-range tasks, and processes key-value data

* Sink:
  
  * A new *TiKV sink* to batch put data into downstream TiKV cluster, with configurable concurrency, batch size, and backoff parameters.

## Prototype

We have developed a [prototype](https://github.com/pingyu/tikv/issues/1) to verify feasibility of this proposal. All of the correctness validation cases are passed. And the benchmark results are as follows:

> | Case                          | gPRC P99 Duration (ms) | gPRC Avg Duration (ms) |

Reviewer 3: should we update the test result based on encoding timestamp into the key part?

Author: Ah yes. We will update the result when we finish the benchmark this week.

![rawkv-cdc](../media/rawkv-cdc.png)

### 2. Add Timestamp to Data

As a kind of [Change Data Capture](https://en.wikipedia.org/wiki/Change_data_capture), timestamp (or version) is necessary to indicate which and when data is changed.

#### 2.1 Requirement

Among requests of a key, the timestamp must be monotonic with the sequence of data flushed to disk in TiKV.

In general, if request `a` ["happened before"](https://en.wikipedia.org/wiki/Happened-before) `b`, then `Timestamp(a) < Timestamp(b)`. As to cross cluster replication, we provide [Causal Consistency](https://en.wikipedia.org/wiki/Causal_consistency) by keeping order of the timestamp the same as sequence of data flush to disk in TiKV. Downstream systems apply data according to the timestamp order, will always lead to the consistent result.

At the same time, as RawKV doesn't provide cross-rows transaction and snapshot isolation, we treat requests of different keys as concurrent, to reduce complexity and improve performance.

#### 2.2 Timestamp Generation

Timestamp is generated by TiKV internally, to get a better overall performance and client compatibility.

> We use [HLC](https://cse.buffalo.edu/tech-reports/2014-04.pdf) method to generate timestamp, which not only is easy to use (by being closed to real world time), but also can capture causality.

Reviewer 3: could you add move info about how many bits are physical and logical timestamp occupies?

Author: OK. Done, PTAL~

##### 2.2.1 Physical Part of Timestamp

Physical part of HLC is acquired from TSO of [PD](https://github.com/tikv/pd).

TSO is a global monotonically increasing timestamp, which help the generation of timestamp be independent to local clock of machine, and be immune to issues such as reverse between machine reboot.

> Physical part is refreshed by repeatedly acquiring TSO in a period of `500ms`, to keep it being closed the real world time.

Reviewer 3: is the period `500ms` configurable?

Author: Yes. I add more words for this, PTAL~

On startup, physical part must be initialized by a successful TSO, to ensure that HLC is larger than last running. RawKV write operations will fail if the initialization has not been completed. (TiKV also can't startup when PD is out of service).

After that, HLC can tolerate fault of TSO. RawKV operations are normal, but time-related metrics such as RPO is not reasonable. Meanwhile, as TiKV-CDC being highly dependent to PD (for cluster and task management, checkpoint, etc.), the replication will not be working.

##### 2.2.2 Logical Part of Timestamp

Logical part is advanced on every write request.

> Moreover, the logical part is advanced on *Leader Transfer*. As TiKV is a distributed system, every store in the TiKV cluster has an instance of timestamp generator. As every key exists in only one region and only one peer (the leader) in the region can write, it's safe to generate timestamps locally in each store, except for leader transfer.

Reviewer 3: > the logical part is advanced on *Leader Transfer*

when assigning a timestamp for a key-value put by a client, the physical part is generated by the local store's machine time, but how to get the logical part? I'm a little confused about this sentence, could you explain with some examples?

Author: I add a diagram and some words for this. PTAL~

On leader transfer, the timestamp generated by the original store could be larger than the store where the new leader is located, which will violate the causal consistency requirement. We solve this by this method: Every peer observes the raft message applied and keeps the maximum timestamp of the region (called `Region-Max-TS`). To become leader, this peer advances the timestamp generator of its store to be larger than `Region-Max-TS`. 

Besides, other region changes should be carefully handled:

* On region merge: The `Region-Max-TS` of the region is the max of the two original regions.

* On region split: The `Region-Max-TS` of the new region is copied from the original one.

#### 2.3 Resolved Timestamp

**TiKV-CDC** using *Resolved Timestamp* (or `resolved-ts`) to indicate the largest transaction timestamp of the replication task, to help downstream avoid getting partial data of a transaction. Downstream streaming system can also use `resolved-ts` as [*watermark*](https://www.oreilly.com/radar/the-world-beyond-batch-streaming-102/).

> As no transaction support, in RawKV scenario we don't need scanning locks in *lock_cf*. But we should notify `resolved-ts` event generation that the minimum timestamp of data that on the way from *proposed* to *applied* (by [`resolver.track_lock`](https://github.com/tikv/tikv/blob/v5.0.4-20211201/components/resolved_ts/src/resolver.rs#L47-L57)).

Reviewer 3: > But we should notify `resolved-ts` event generation that the minimum timestamp of data that on the way from *proposed* to *applied* (by [`resolver.track_lock`](https://github.com/tikv/tikv/blob/v5.0.4-20211201/components/resolved_ts/src/resolver.rs#L47-L57)).

I'm a little confused, does it mean we need to record the minimal timestamp of data that is proposed but not applied?

Author: Yes. I add a diagram and improve some words, PTAL~

### 3. Deletion

To make deletions be captured as change data, we turn physical deletion to logical deletion, i.e., set a *deleted* flag, other than physically delete the entry.

> The *garbage collection* of deleted data is implemented by setting TTL to 25 hours after, and *GC-ed* by compaction filter. At the same time, start timestamp (i.e. `start-ts`) of **TiKV-CDC** replication task should not be earlier than 24 hours before.

Reviewer 3: > The *garbage collection* of deleted data is implemented by setting TTL to 25 hours after.

could you add some explanation on why setting it to 25 hours instead of 1 second?

Author: Sure. Done, PTAL~

TSO is a global monotonically increasing timestamp, which help the generation of timestamp be independent to local clock of machine, and be immune to issues such as reverse between machine reboot.

Physical part is refreshed by repeatedly acquiring TSO in a period of `500ms`, to keep it being closed the real world time.

On startup, physical part must be initialized by a successful TSO, to ensure that HLC is larger than last running. RawKV write operations will fail if the initialization has not been completed. (TiKV also can't startup when PD is out of service).

> After that, HLC can tolerate fault of TSO. RawKV operations are normal, but time-related metrics such as RPO is not reasonable. Meanwhile, as TiKV-CDC being highly dependent to PD (for cluster and task management, checkpoint, etc.), the replication will not be working.

Reviewer 2: What exactly is the "After **that**"? From the context, it is the TSO failure on startup but I think it should be TSO failure after startup. And I think the time-related metrics may not be affected unless the clock shift is crazy enough...

Author: It is "After successful initialization". I'm sorry for the not clear enough.

We do not advance HLC by local clock when TSO is offline, which will cause the time-related metrics problem.

Using local clock on fault of TSO need some other logics to ensure causality, such as saving maximum HLC on storage to avoid timestamp reverse between startup.

I do not include the "Falling Back to Local Clock" in this proposal for simplicity, as the PD offline is rare, and when it was offline it would cause more severe issues than the problem of time-related metrics.

Reviewer 2: > Using local clock on fault of TSO need some other logics to ensure causality, such as saving maximum HLC on storage to avoid timestamp reverse between startup.

If I understand it right, the `RegionMaxTs` and `StoreMaxTs` should contain both physical time and logical time, thus we always know what is the latest physical time and therefore safe to use the local clock.

Author: Yes it is safe during running.
But unsafe will happen on restart. If the local clock shift backward between TiKV stop & restart, it's possible to get a smaller timestamp than last running.

Author: CockroachDB has a description on this restart issue [here](https://github.com/cockroachdb/cockroach/blob/d5ee674ec410f384df9ffb6f737e9b4e2c922589/pkg/util/hlc/doc.go#L152-L160). We can refer to it.

Reviewer 2: > Yes it is safe during running.

Yes, that's my point. As long as the TiKV doesn't restart, the metrics of RPO and write operation should not affect by PD failure.

Reviewer 3: it would be better if you can explain the reason with an example. For example:

In a 3 AZ deployment:
- PD leader is located in AZ1
- TiKV region leaders are uniformly distributed in AZ1 - AZ3
- There is a partial network partition that happens: AZ1 can not communicate with AZ2, but they can both communicate with AZ3, TiKV Client (for example, client-java) can connect to all the three AZs

Would the write requests to AZ2 TiKV Region be blocked because AZ2 can not sync physical timestamp with PD leader in AZ1?

Author: I improve the words here. PTAL~
